orttraining/orttraining/training_api/optimizer.cc

```diff
@@ -40,53 +49,93 @@ Status Optimizer::ConstructInputs() {
+    std::vector<Tensor> params, first_order_moments, second_order_moments;
+    // TODO: Change to tensor seq implementation once clip grad norm op
+    // that accepts tensor seq as input for gradients is complete.
+    std::vector<OrtValue> grads;
+
+    // Collect all the params, grads, and moments based on grad names
+    // Input names 0-4 are reserved for lr, step, params, first order moments, second order moments
+    // input names 5 onwards are all the gradient names.
+    // Collect all the inputs based on the gradient names order.
+    for (size_t i = 5; i < input_names_.size(); i++) {
+      std::string param_name;
+      if (utils::GetParamNameFromGradient(input_names_[i], param_name)) {
+        const auto named_parameter_it = named_parameters_.find(param_name);
+        ORT_ENFORCE(named_parameter_it != named_parameters_.end(),
+                    "Unknown param: ", param_name, " for field: ", input_names_[i]);
+
+        // collect the gradients as ortvalues
+        grads.push_back(named_parameter_it->second->Gradient());
+
+        // collect parameters and prepare for tensorseq creation
+        auto* param_tensor = named_parameter_it->second->Data().GetMutable<Tensor>();
+        params.emplace_back(
+            Tensor(param_tensor->DataType(), param_tensor->Shape(),
+                   param_tensor->MutableDataRaw(), param_tensor->Location()));
+
+        // collect first order moments and prepare for tensorseq creation
+        auto* first_order_moment_tensor = param_named_optimizer_states.at(param_name)
+                                              .momentum_named_states.at(MOMENT_STATE_NAMES[0])
+                                              .GetMutable<Tensor>();
+        first_order_moments.emplace_back(
+            Tensor(first_order_moment_tensor->DataType(), first_order_moment_tensor->Shape(),
+                   first_order_moment_tensor->MutableDataRaw(), first_order_moment_tensor->Location()));
+
+        // collect second order moments and prepare for tensorseq creation
+        auto* second_order_moment_tensor = param_named_optimizer_states.at(param_name)
+                                               .momentum_named_states.at(MOMENT_STATE_NAMES[1])
+                                               .GetMutable<Tensor>();
+        second_order_moments.emplace_back(
+            Tensor(second_order_moment_tensor->DataType(), second_order_moment_tensor->Shape(),
+                   second_order_moment_tensor->MutableDataRaw(), second_order_moment_tensor->Location()));
       } else {
-        ORT_ENFORCE("This is an invalid graph. Optimizer graph contains unknown user input:", name);
+        ORT_ENFORCE(
+            false, "This is an invalid graph. Optimizer graph contains unknown user input:", input_names_[i]);
       }
     }
+
+    const auto tensorseq_inserter = [](auto& tensors, auto* inputs) {
```

Comment:
I believe this lamda can be put outside of loop.

Reply:
The lambda is not inside any for loop. It is at the function body level.